src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets

```diff
@@ -1529,29 +1531,37 @@ because xbuild doesn't support framework reference assemblies.
   </ItemGroup>
 
   <GenerateJavaStubs
-      AndroidRuntime="$(_AndroidRuntime)"
       CodeGenerationTarget="$(_AndroidJcwCodegenTarget)"
       ResolvedAssemblies="@(_ResolvedAssemblies)"
       ResolvedUserAssemblies="@(_ResolvedUserMonoAndroidAssemblies)"
       ErrorOnCustomJavaObject="$(AndroidErrorOnCustomJavaObject)"
       Debug="$(AndroidIncludeDebugSymbols)"
       AndroidSdkPlatform="$(_AndroidApiLevel)"
       OutputDirectory="$(IntermediateOutputPath)android"
-      TypemapOutputDirectory="$(_NativeAssemblySourceDir)"
-      GenerateNativeAssembly="$(_AndroidGenerateNativeAssembly)"
       PackageNamingPolicy="$(AndroidPackageNamingPolicy)"
       ApplicationJavaClass="$(AndroidApplicationJavaClass)"
       FrameworkDirectories="$(_XATargetFrameworkDirectories);$(_XATargetFrameworkDirectories)Facades"
       SupportedAbis="@(_BuildTargetAbis)"
-      SkipJniAddNativeMethodRegistrationAttributeScan="$(_SkipJniAddNativeMethodRegistrationAttributeScan)"
       EnableMarshalMethods="$(_AndroidUseMarshalMethods)"
-      EnableManagedMarshalMethodsLookup="$(_AndroidUseManagedMarshalMethodsLookup)"
-      LinkingEnabled="$(_LinkingEnabled)"
-      HaveMultipleRIDs="$(_HaveMultipleRIDs)"
-      IntermediateOutputDirectory="$(IntermediateOutputPath)"
-      Environments="@(_EnvironmentFiles)">
+      IntermediateOutputDirectory="$(IntermediateOutputPath)">
   </GenerateJavaStubs>
 
+  <RewriteMarshalMethods
+      Debug="$(AndroidIncludeDebugSymbols)"
+      EnableManagedMarshalMethodsLookup="$(_AndroidUseManagedMarshalMethodsLookup)"
+      EnableMarshalMethods="$(_AndroidUseMarshalMethods)"
+      Environments="@(_EnvironmentFiles)"
+      IntermediateOutputDirectory="$(IntermediateOutputPath)">
+  </RewriteMarshalMethods>
+
+  <GenerateTypeMappings
+      AndroidRuntime="$(_AndroidRuntime)"
+      Debug="$(AndroidIncludeDebugSymbols)"
+      GenerateNativeAssembly="$(_AndroidGenerateNativeAssembly)"
+      IntermediateOutputDirectory="$(IntermediateOutputPath)"
+      SkipJniAddNativeMethodRegistrationAttributeScan="$(_SkipJniAddNativeMethodRegistrationAttributeScan)"
+      TypemapOutputDirectory="$(_NativeAssemblySourceDir)">
+  </GenerateTypeMappings>
```

Comment:
Can the whole task be skipped in some cases?

```csharp
		bool useMarshalMethods = !Debug && EnableMarshalMethods;

		if (!useMarshalMethods)
			return true;
```

I think it would be clearer if a `Condition` did this part, since it's refactored into one task now. 👍 

The `Condition` would also be logged in a build log, where the existing code doesn't log.